prac_07/project_management.py

```diff
@@ -0,0 +1,216 @@
+    return number
+
+
+def get_valid_percentage(prompt):
+    is_valid_input = False
+    while not is_valid_input:
+        try:
+            percentage = int(input(prompt).strip())
+            if percentage < 0 or percentage > 100:
+                print("Invalid percentage, must be between 0 and 100")
+            else:
+                is_valid_input = True
+        except ValueError:
+            print("Invalid input (not an integer)")
+    return percentage
+
+
+def save_file(projects):
+    filename = input("Would you like to save to projects.txt? ").strip()
+    if filename != "":
+        save_projects(filename, projects)
+    else:
+        filename = DEFAULT_FILENAME
+        save_projects(filename, projects)
+
+
+def load_new_file():
+    filename = input("filename (default - projects.txt): ").strip()
+    if filename != "":
+        projects = load_projects(filename)
+    else:
+        filename = DEFAULT_FILENAME
+        projects = load_projects(filename)
+    print(f"Loaded {len(projects)} from {filename}")
+    return projects
+
+
+def display_project_completion_status(projects):
+    """Display all projects based on completion status"""
+    sorted_projects = sorted(projects, key=attrgetter("priority"))
+    incomplete_projects = [project for project in sorted_projects if project.is_incomplete()]
+    complete_projects = [project for project in sorted_projects if project.is_complete()]
+    print("Incomplete projects:")
+    display_projects(incomplete_projects)
+    print("Complete projects:")
+    display_projects(complete_projects)
+
+
+def display_projects(projects):
+    """Display projects."""
+    for project in projects:
+        print(f"  {project}")
+
+
+def add_new_project(projects):
+    """Add new project object to list."""
+    print("Let's add a new project")
+    name = get_valid_string("Name: ")
+    start_date = get_valid_date("Start date (dd/mm/yyyy): ")
+    # Priority can't be lower than 1
+    priority = get_valid_number("Priority: ", 1)
+    # Cost can't be lower than $0
+    cost_estimate = float(get_valid_number("Cost estimate: $", 0))
+    completion_percentage = get_valid_percentage("Percentage complete: ")
+    new_project = Project(name, start_date.strftime("%d/%m/%Y"), priority, cost_estimate, completion_percentage)
+    projects.append(new_project)
+
+
+def get_valid_string(prompt):
+    string = input(prompt).strip()
+    while string == "":
+        print("Invalid string must not be empty")
+        string = input(prompt)
+    return string
+
+
+def get_valid_date(prompt):
+    is_valid_input = False
+    while not is_valid_input:
+        try:
+            date = input(prompt)
+            date = convert_to_datetime_object(date)
+            is_valid_input = True
+        except ValueError:
+            print("Invalid date")
+    return date
+
+
+def convert_to_datetime_object(string):
+    date = datetime.datetime.strptime(string, "%d/%m/%Y").date()
+    return date
```

Comment:
MD, And I would check with lindsay as to whether this needed to be its own function